src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,755 @@
+            }
+
+            return leftMost;
+        }
+    }
+
+    final static class GatherSink<T, A, R> implements Sink<T>, Gatherer.Downstream<R> {
+        private final Sink<R> sink;
+        private final Gatherer<T, A, R> gatherer;
+        private final Integrator<A, T, R> integrator; // Optimization: reuse
+        private A state;
+        private boolean proceed = true;
+
+        GatherSink(Gatherer<T, A, R> gatherer, Sink<R> sink) {
+            this.gatherer = gatherer;
+            this.sink = sink;
+            this.integrator = gatherer.integrator();
+        }
+
+        // java.util.stream.Sink contract below:
+
+        @Override
+        public void begin(long size) {
+            final var initializer = gatherer.initializer();
+            if (initializer != Gatherer.defaultInitializer()) // Optimization
+                state = initializer.get();
+            sink.begin(size);
+        }
+
+        @Override
+        public void accept(T t) {
+            /* Benchmarks have indicated that doing an unconditional write to
+             * `proceed` is more efficient than branching.
+             * We use `&=` here to prevent flips from `false` -> `true`.
+             *
+             * As of writing this, taking `greedy` or `stateless` into
+             * consideration at this point doesn't yield any performance gains.
+             */
+            proceed &= integrator.integrate(state, t, this);
+        }
+
+        @Override
+        public boolean cancellationRequested() {
+            return cancellationRequested(proceed);
+        }
+
+        private boolean cancellationRequested(boolean knownProceed) {
+            // Highly performance sensitive
+            return !(knownProceed && (!sink.cancellationRequested() || (proceed = false)));
+        }
+
+        @Override
+        public void end() {
+            final var finisher = gatherer.finisher();
+            if (finisher != Gatherer.<A, R>defaultFinisher()) // Optimization
+                finisher.accept(state, this);
+            sink.end();
+            state = null; // GC assistance
+        }
+
+        // Gatherer.Sink contract below:
+
+        @Override
+        public boolean isRejecting() {
+            return !proceed;
+        }
+
+        @Override
+        public boolean push(R r) {
+            var p = proceed;
+            if (p)
+                sink.accept(r);
+            return !cancellationRequested(p);
+        }
+    }
+
+    private static int opFlagsFor(Integrator<?, ?, ?> integrator) {
+        return integrator instanceof Integrator.Greedy<?, ?, ?>
+                ? GREEDY_FLAGS : SHORT_CIRCUIT_FLAGS;
+    }
+
+    private final static int DEFAULT_FLAGS =
+            StreamOpFlag.NOT_SORTED | StreamOpFlag.NOT_DISTINCT |
+                    StreamOpFlag.NOT_SIZED;
+
+    private final static int SHORT_CIRCUIT_FLAGS =
+            DEFAULT_FLAGS | StreamOpFlag.IS_SHORT_CIRCUIT;
+
+    private final static int GREEDY_FLAGS =
```

Comment:
Let's use blessed modifiers order:
```suggestion
    private static final int DEFAULT_FLAGS =
            StreamOpFlag.NOT_SORTED | StreamOpFlag.NOT_DISTINCT |
                    StreamOpFlag.NOT_SIZED;

    private static final int SHORT_CIRCUIT_FLAGS =
            DEFAULT_FLAGS | StreamOpFlag.IS_SHORT_CIRCUIT;

    private static final int GREEDY_FLAGS =
```